dropletutils-read-10x-counts.R

```diff
@@ -34,19 +67,50 @@ option_list = list(
 opt <- wsc_parse_args(option_list, mandatory = c('samples', 'output_object_file'))
 
 # Now we're hapy with the arguments, load DropletUtils and do the work
-
 suppressPackageStartupMessages(require(DropletUtils))
 
 # Parse out multiple directories from the samples string
-
 samples <- wsc_split_string(opt$samples)
 
 # Derive the SingleCellExperiment
+single_cell_experiment <- read10xCounts( samples = samples, col.names = opt$col_names )
+
+# metadata file logic 
+if(!is.na(opt$metadata_files)){
+  metadata_files = wsc_split_string(opt$metadata_files)
+  # check cell id column is provided
+  if(is.na(opt$cell_id_column)){
+    stop("Cell id column name for SDRF files is not provided. ")
+  }
+  metadata = lapply(metadata_files, function(x) read.csv(x, sep="\t"))
+  metadata = do.call(rbind, metadata)
+  # remove technical duplicate rows
+  metadata = metadata[which(!duplicated(metadata[, opt$cell_id_column])), ]
 
-single_cell_experiment <- read10xCounts( samples = samples, col.names = opt$col_names  )
+  # keep only selected columns, if specified 
+  if(!is.na(opt$metadata_columns)){
+    metadata_columns = wsc_split_string(opt$metadata_columns)
+    if(!all(metadata_columns %in% colnames(metadata))){
+      stop("Provided metadata column names do not match existing names.")
+    } 
+    # subset by given columns
+    metadata = metadata[, metadata_columns]
+  }
 
-# Print an object summary
```

Comment:
I think the below is too complex. You already removed potential duplicates so, so you just need to check that all cells have metadata, then do a simple merge() without sort:

```
if (! all(metadata[[opt$cell_id_column]] %in% colData(single_cell_experiment$Barcode))){
  stop("Error in matching metadata to expression matrix columns.")
}else{
  colData(single_cell_experiment) <- merge(colData(single_cell_experiment), metadata, by.x = 'Barcode', by.y = opt$cell_id_column, sort = FALSE)
}
```